k4MarlinWrapper/src/components/EDM4hep2Lcio.cpp

```diff
@@ -386,25 +386,33 @@ void EDM4hep2LcioTool::convertAdd(const std::string& e4h_coll_name, const std::s
 // Use the collection names in the parameters to read and write them
 StatusCode EDM4hep2LcioTool::convertCollections(lcio::LCEventImpl* lcio_event) {
   std::optional<std::reference_wrapper<const podio::Frame>> edmEvent;
-  if (m_collectionNames.empty() && m_podioDataSvc) {
-    edmEvent          = m_podioDataSvc->getEventFrame();
-    m_collectionNames = edmEvent.value().get().getAvailableCollections();
-  } else if (m_collectionNames.empty()) {
-    std::optional<std::map<uint32_t, std::string>> idToNameOpt(std::move(m_idToName));
-    auto                                           collections = getAvailableCollectionsFromStore(this, idToNameOpt);
-    m_idToName                                                 = std::move(idToNameOpt.value());
-    m_collectionNames.insert(m_collectionNames.end(), collections.begin(), collections.end());
-  }
-  // Start off with the pre-defined collection name mappings
-  auto collsToConvert{m_collNames.value()};
-  // We *always* want to convert the EventHeader
-  collsToConvert.emplace(edm4hep::labels::EventHeader, "<directly into LCEvent>");
-  if (m_convertAll) {
-    info() << "Converting all collections from EDM4hep to LCIO" << endmsg;
-    // And simply add the rest, exploiting the fact that emplace will not
-    // replace existing entries with the same key
-    for (const auto& name : m_collectionNames) {
-      collsToConvert.emplace(name, name);
+  // use m_collsToConvert to detect whether we run the first time and cache the
+  // results as we can assume that all the events have the same contents
+  if (m_collsToConvert.empty()) {
+    // Start off with the pre-defined collection name mappings
+    m_collsToConvert = {m_collNames.value().begin(), m_collNames.value().end()};
+    // We *always* want to convert the EventHeader
+    m_collsToConvert.emplace(edm4hep::labels::EventHeader, "<directly into LCEvent>");
+
+    if (m_convertAll) {
+      info() << "Converting all collections from EDM4hep to LCIO" << endmsg;
+      std::vector<std::string> collectionNames{};
+      if (m_podioDataSvc) {
+        // If we have the PodioDataSvc get the collections available from frame
+        edmEvent        = m_podioDataSvc->getEventFrame();
+        collectionNames = edmEvent.value().get().getAvailableCollections();
+      }
+      // Always check the contents of the TES
```

Comment:
Always because when using the PodioDataSvc with functional algorithms the collections don't go to the event frame, otherwise this would be an `if-else` (I thought initially it should be an `if-else`).

Reply:
Yes. Should I put this info into the comment?